conversion/conversion.go

```diff
@@ -445,26 +440,60 @@ func UpdateDDLForeignKeys(project, instance, dbName string, conv *internal.Conv,
-	for i, fkStmt := range fkStmts {
-		// TODO: Improve performance of the foreign key constraint updates.
-		// For example, issue all of the update ops first before waiting for them to complete
-		// so that that can execute in parallel. We could also print out ids of the
-		// long-running operations.
-		op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
-			Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
-			Statements: []string{fkStmt},
-		})
-		if err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
-			continue
-		}
-		if err := op.Wait(ctx); err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+	type Pair struct {
+		a, b interface{}
+	}
+	fkStmtsQueue := make(chan string, len(fkStmts))
+	results := make(chan Pair, len(fkStmts))
+
+	for _, fkStmt := range fkStmts {
+		fkStmtsQueue <- fkStmt
+	}
+	close(fkStmtsQueue)
+
+	// Spawn `maxWorkers` go routines that each read one fkStatement from the fkStmtsQueue channel.
+	// On succesful creation of a foreign key, push a Pair to the results channel where Pair.b is 1
+	// on success and 0 on failure.
+	for i := int64(1); i <= maxWorkers; i++ {
+		go func(fkStmtsQueue <-chan string, results chan<- Pair) {
+
+			for fkStmt := range fkStmtsQueue {
+				adminClient, err := database.NewDatabaseAdminClient(ctx)
+				if err != nil {
+					fmt.Printf("can't create admin client: %s\n", analyzeError(err, project, instance))
+					conv.Unexpected(fmt.Sprintf("can't create admin client: %s\n", analyzeError(err, project, instance)))
+					results <- Pair{fkStmt, 0}
+					continue
+				}
+				defer adminClient.Close()
+				op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
+					Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
+					Statements: []string{fkStmt},
+				})
+				if err != nil {
+					fmt.Printf("Cannot submit request for create foreign key with statement %s: %s\n", fkStmt, err)
+					conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+					results <- Pair{fkStmt, 0}
+					continue
+				}
+				if err := op.Wait(ctx); err != nil {
+					fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
+					conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+					results <- Pair{fkStmt, 0}
+					continue
+				}
+				internal.VerbosePrintln("Updated schema with statement: " + fkStmt)
+				results <- Pair{fkStmt, 1}
+			}
+		}(fkStmtsQueue, results)
+	}
+	progress := int64(0)
+	for i := 1; i <= len(fkStmts); i++ {
+		pair := <-results
+		if pair.b == 1 {
```

Comment:
why do we report progress only for successful foreign key schema changes?

Reply:
The errors are printed out inside the go routine on failure before exiting while the thread is running.

Reply:
Also we do a p.Done() at the end which automatically updates progress to 100%.